File: langs/nodejs/forgeh/src/engine.js
// ForgeHash-B3 v1 — core engine: seed derivation, memory initialization,
// memory filling, and finalization. Implements SPECIFICATION.md §§9-17.
//
// This is a deliberately single-threaded reference implementation, matching
// the C#/Rust reference engines word-for-word.

import * as blake3 from "blake3";
import {
  BLOCK_SIZE,
  WORDS,
  mix,
  bytesToWords,
  wordsToBytes,
  xorWords,
  rotl,
} from "./forgemix.js";
import { MAX_PASSWORD } from "./params.js";

const SEED_CTX = "ForgeHash/v1/seed";
const EXPAND_PREFIX = Buffer.from("ForgeHash/v1/expand", "utf8");
const GROUP_PREFIX = Buffer.from("ForgeHash/v1/group", "utf8");
const GROUP_ROOT_PREFIX = Buffer.from("ForgeHash/v1/group-root", "utf8");
const FINAL_PREFIX = Buffer.from("ForgeHash/v1/final", "utf8");
const OUTPUT_PREFIX = Buffer.from("ForgeHash/v1/output", "utf8");

const MASK64 = (1n << 64n) - 1n;

function le32(v) {
  const b = Buffer.alloc(4);
  b.writeUInt32LE(v >>> 0, 0);
  return b;
}

function le64(v) {
  const b = Buffer.alloc(8);
  b.writeBigUInt64LE(BigInt(v) & MASK64, 0);
  return b;
}

/**
 * Build the little-endian input buffer described in SPECIFICATION.md §9:
 *
 *   LE32(version=1) || LE32(memoryKiB) || LE32(iterations) || LE32(parallelism)
 *   || LE32(outputLength) || LE64(passwordLength) || password
 *   || LE32(saltLength) || salt || LE32(contextLength=0)
 */
function buildEncodedInput(password, salt, params) {
  return Buffer.concat([
    le32(1),
    le32(params.memoryKiB),
    le32(params.iterations),
    le32(params.parallelism),
    le32(params.outputLength),
    le64(password.length),
    password,
    le32(salt.length),
    salt,
    le32(0), // empty context, always present
  ]);
}

function deriveSeedBytes(material) {
  return blake3.deriveKey(SEED_CTX, material, { length: 32 });
}

/** Expand(input, outputLength) = BLAKE3-XOF("ForgeHash/v1/expand" || input) */
function expand(input, outLen) {
  const hasher = blake3.createHash();
  hasher.update(EXPAND_PREFIX);
  hasher.update(input);
  return hasher.digest(undefined, { length: outLen });
}

function blake3Hash(input) {
  return blake3.hash(input, { length: 32 });
}

function blake3Xof(input, outLen) {
  const hasher = blake3.createHash();
  hasher.update(input);
  return hasher.digest(undefined, { length: outLen });
}

/** FastRange(x, n) = High64(x * n) using a 128-bit intermediate product. */
function fastRange(x, n) {
  return (x * n) >> 64n;
}

class Memory {
  constructor(blockCount, parallelism) {
    this.parallelism = parallelism;
    this.blocksPerLane = blockCount / parallelism;
    this.sliceLength = this.blocksPerLane / 4;
    this.words = new Array(blockCount * WORDS).fill(0n);
  }

  flatOffset(lane, index) {
    return (lane * this.blocksPerLane + index) * WORDS;
  }

  getBlock(lane, index, out) {
    const start = this.flatOffset(lane, index);
    const dest = out || new Array(WORDS);
    for (let i = 0; i < WORDS; i++) dest[i] = this.words[start + i];
    return dest;
  }

  setBlock(lane, index, block) {
    const start = this.flatOffset(lane, index);
    for (let i = 0; i < WORDS; i++) this.words[start + i] = block[i];
  }
}

/**
 * Derive the ForgeHash initial seed (32 bytes) from a password, salt, and
 * parameters. Throws if the password or parameters are invalid.
 */
export function deriveSeed(password, salt, params) {
  if (password.length > MAX_PASSWORD) {
    throw new RangeError("ForgeHash: password too long");
  }
  params.validate();
  if (salt.length < 16 || salt.length > 64) {
    throw new RangeError("ForgeHash: salt length out of range");
  }
  const material = buildEncodedInput(password, salt, params);
  return deriveSeedBytes(material);
}

/**
 * Derive the full ForgeHash-B3 output (length `params.outputLength` bytes)
 * from a password, salt, and parameters. This performs the full
 * memory-hard computation and may take a significant amount of time and
 * memory depending on `params`.
 */
export function deriveHash(password, salt, params) {
  if (password.length > MAX_PASSWORD) {
    throw new RangeError("ForgeHash: password too long");
  }
  params.validate();
  if (salt.length < 16 || salt.length > 64) {
    throw new RangeError("ForgeHash: salt length out of range");
  }

  const material = buildEncodedInput(password, salt, params);
  const seed = deriveSeedBytes(material);

  const memory = new Memory(params.memoryKiB, params.parallelism);
  initializeMemory(memory, seed);
  fillMemory(memory, params.iterations);
  return finalize(memory, seed, params);
}

function initializeMemory(memory, seed) {
  for (let lane = 0; lane < memory.parallelism; lane++) {
    for (let blockIndex = 0; blockIndex < 2; blockIndex++) {
      const input = Buffer.concat([seed, le32(lane), le32(blockIndex)]);
      const blockBytes = expand(input, BLOCK_SIZE);
      const words = bytesToWords(blockBytes);
      memory.setBlock(lane, blockIndex, words);
    }
  }
}

function fillMemory(memory, iterations) {
  for (let pass = 0; pass < iterations; pass++) {
    for (let slice = 0; slice < 4; slice++) {
      for (let lane = 0; lane < memory.parallelism; lane++) {
        processSlice(memory, pass, slice, lane);
      }
    }
  }
}

// Scratch buffers reused across block iterations to avoid per-block allocation.
const scratchPrevious = new Array(WORDS);
const scratchReference = new Array(WORDS);
const scratchCombined = new Array(WORDS);
const scratchMixed = new Array(WORDS);
const scratchOldCurrent = new Array(WORDS);
const scratchOutput = new Array(WORDS);

function processSlice(memory, pass, slice, lane) {
  let start = slice * memory.sliceLength;
  const end = start + memory.sliceLength;
  if (pass === 0 && slice === 0) {
    start = 2;
  }

  for (let blockIndex = start; blockIndex < end; blockIndex++) {
    const previousIndex =
      blockIndex > 0 ? blockIndex - 1 : memory.blocksPerLane - 1;
    memory.getBlock(lane, previousIndex, scratchPrevious);

    const { referenceLane, referenceIndex } = selectReference(
      memory,
      pass,
      slice,
      lane,
      blockIndex,
      scratchPrevious
    );
    memory.getBlock(referenceLane, referenceIndex, scratchReference);
    xorWords(scratchPrevious, scratchReference, scratchCombined);

    const passBig = BigInt(pass);
    const laneBig = BigInt(lane);
    const blockIndexBig = BigInt(blockIndex);

    if (pass === 0) {
      mix(scratchCombined, passBig, laneBig, blockIndexBig, scratchOutput);
      memory.setBlock(lane, blockIndex, scratchOutput);
    } else {
      mix(scratchCombined, passBig, laneBig, blockIndexBig, scratchMixed);
      memory.getBlock(lane, blockIndex, scratchOldCurrent);
      xorWords(scratchOldCurrent, scratchMixed, scratchOutput);
      memory.setBlock(lane, blockIndex, scratchOutput);
    }
  }
}

function selectReference(memory, pass, slice, currentLane, blockIndex, previous) {
  const addressWord =
    (previous[0] ^
      rotl(previous[17], 13) ^
      previous[73] ^
      BigInt(pass) ^
      rotl(BigInt(blockIndex), 29)) &
    MASK64;

  let referenceLane;
  if (memory.parallelism === 1) {
    referenceLane = 0;
  } else if (blockIndex % 32 === 0) {
    referenceLane = Number(fastRange(previous[1], BigInt(memory.parallelism)));
  } else {
    referenceLane = currentLane;
  }

  let allowed = allowedBlockCount(
    memory,
    pass,
    slice,
    currentLane,
    referenceLane,
    blockIndex
  );
  if (allowed === 0) {
    referenceLane = currentLane;
    allowed = allowedBlockCount(
      memory,
      pass,
      slice,
      currentLane,
      referenceLane,
      blockIndex
    );
  }
  const referenceIndex = Number(fastRange(addressWord, BigInt(allowed)));
  return { referenceLane, referenceIndex };
}

function allowedBlockCount(memory, pass, slice, currentLane, referenceLane, blockIndex) {
  if (referenceLane === currentLane) {
    return pass === 0 ? blockIndex : memory.blocksPerLane;
  }
  return slice * memory.sliceLength;
}

function finalize(memory, seed, params) {
  const acc = new Array(WORDS).fill(0n);
  const last = memory.blocksPerLane - 1;
  const quarter = Math.floor(memory.blocksPerLane / 4);
  const half = Math.floor(memory.blocksPerLane / 2);
  const threeQuarter = Math.floor((memory.blocksPerLane * 3) / 4);
  const tmp = new Array(WORDS);

  for (let lane = 0; lane < memory.parallelism; lane++) {
    for (const index of [last, quarter, half, threeQuarter]) {
      memory.getBlock(lane, index, tmp);
      for (let i = 0; i < WORDS; i++) acc[i] ^= tmp[i];
    }
  }

  const accumulatorBytes = wordsToBytes(acc);
  const groupRoot = computeGroupRoot(memory);

  const rootInput = Buffer.concat([
    FINAL_PREFIX,
    seed,
    accumulatorBytes,
    groupRoot,
    le32(params.memoryKiB),
    le32(params.iterations),
    le32(params.parallelism),
    le32(params.outputLength),
  ]);
  const root = blake3Hash(rootInput);

  const outputInput = Buffer.concat([OUTPUT_PREFIX, root, seed]);
  return blake3Xof(outputInput, params.outputLength);
}

function computeGroupRoot(memory) {
  const rootHasher = blake3.createHash();
  rootHasher.update(GROUP_ROOT_PREFIX);

  const GROUP = 64;
  const total = memory.parallelism * memory.blocksPerLane;
  const groupCount = Math.ceil(total / GROUP);
  const groupBuf = Buffer.alloc(GROUP * BLOCK_SIZE);
  const words = new Array(WORDS);
  const blockBytes = Buffer.alloc(BLOCK_SIZE);

  for (let groupIndex = 0; groupIndex < groupCount; groupIndex++) {
    const start = groupIndex * GROUP;
    const count = Math.min(total - start, GROUP);
    for (let i = 0; i < count; i++) {
      const flat = start + i;
      const lane = Math.floor(flat / memory.blocksPerLane);
      const blockIndex = flat % memory.blocksPerLane;
      memory.getBlock(lane, blockIndex, words);
      wordsToBytes(words, blockBytes);
      blockBytes.copy(groupBuf, i * BLOCK_SIZE);
    }
    const byteLen = count * BLOCK_SIZE;

    const groupHasher = blake3.createHash();
    groupHasher.update(GROUP_PREFIX);
    groupHasher.update(le64(groupIndex));
    groupHasher.update(le64(count));
    groupHasher.update(groupBuf.subarray(0, byteLen));
    const digest = groupHasher.digest(undefined, { length: 32 });

    rootHasher.update(le64(groupIndex));
    rootHasher.update(digest);
  }

  return rootHasher.digest(undefined, { length: 32 });
}

File: langs/rust/forgeh/src/engine.rs
use crate::error::Error;
use crate::forgemix::{
    bytes_to_words, mix, words_to_bytes, xor_words, BLOCK_SIZE, WORDS,
};
use crate::params::{Params, MAX_PASSWORD};

const SEED_CTX: &str = "ForgeHash/v1/seed";
const EXPAND_PREFIX: &str = "ForgeHash/v1/expand";
const GROUP_PREFIX: &str = "ForgeHash/v1/group";
const GROUP_ROOT_PREFIX: &str = "ForgeHash/v1/group-root";
const FINAL_PREFIX: &str = "ForgeHash/v1/final";
const OUTPUT_PREFIX: &str = "ForgeHash/v1/output";

fn le32(v: u32) -> [u8; 4] {
    v.to_le_bytes()
}
fn le64(v: u64) -> [u8; 8] {
    v.to_le_bytes()
}

fn build_encoded_input(password: &[u8], salt: &[u8], params: &Params) -> Vec<u8> {
    let mut buf = Vec::with_capacity(4 * 5 + 8 + password.len() + 4 + salt.len() + 4);
    buf.extend_from_slice(&le32(1));
    buf.extend_from_slice(&le32(params.memory_kib as u32));
    buf.extend_from_slice(&le32(params.iterations as u32));
    buf.extend_from_slice(&le32(params.parallelism as u32));
    buf.extend_from_slice(&le32(params.output_length as u32));
    buf.extend_from_slice(&le64(password.len() as u64));
    buf.extend_from_slice(password);
    buf.extend_from_slice(&le32(salt.len() as u32));
    buf.extend_from_slice(salt);
    buf.extend_from_slice(&le32(0)); // empty context
    buf
}

fn derive_seed_bytes(material: &[u8]) -> [u8; 32] {
    let mut hasher = blake3::Hasher::new_derive_key(SEED_CTX);
    hasher.update(material);
    *hasher.finalize().as_bytes()
}

fn expand(input: &[u8], out: &mut [u8]) {
    let mut hasher = blake3::Hasher::new();
    hasher.update(EXPAND_PREFIX.as_bytes());
    hasher.update(input);
    hasher.finalize_xof().fill(out);
}

fn blake3_hash(input: &[u8]) -> [u8; 32] {
    *blake3::hash(input).as_bytes()
}

fn blake3_xof(input: &[u8], out_len: usize) -> Vec<u8> {
    let mut hasher = blake3::Hasher::new();
    hasher.update(input);
    let mut out = vec![0u8; out_len];
    hasher.finalize_xof().fill(&mut out);
    out
}

fn fast_range(x: u64, n: u64) -> u64 {
    ((x as u128) * (n as u128) >> 64) as u64
}

fn rotl(x: u64, n: u32) -> u64 {
    x.rotate_left(n)
}

struct Memory {
    words: Vec<u64>,
    parallelism: usize,
    blocks_per_lane: usize,
    slice_length: usize,
}

impl Memory {
    fn new(block_count: usize, parallelism: usize) -> Self {
        let blocks_per_lane = block_count / parallelism;
        Self {
            words: vec![0u64; block_count * WORDS],
            parallelism,
            blocks_per_lane,
            slice_length: blocks_per_lane / 4,
        }
    }

    fn block_mut(&mut self, lane: usize, index: usize) -> &mut [u64] {
        let flat = lane * self.blocks_per_lane + index;
        let start = flat * WORDS;
        &mut self.words[start..start + WORDS]
    }

    fn block(&self, lane: usize, index: usize) -> &[u64] {
        let flat = lane * self.blocks_per_lane + index;
        let start = flat * WORDS;
        &self.words[start..start + WORDS]
    }

    fn copy_block(&self, lane: usize, index: usize, dest: &mut [u64; WORDS]) {
        dest.copy_from_slice(self.block(lane, index));
    }
}

pub fn derive_seed(password: &[u8], salt: &[u8], params: &Params) -> Result<[u8; 32], Error> {
    if password.len() > MAX_PASSWORD {
        return Err(Error::Params("password too long"));
    }
    params.validate()?;
    if salt.len() < 16 || salt.len() > 64 {
        return Err(Error::Params("salt length out of range"));
    }
    let material = build_encoded_input(password, salt, params);
    Ok(derive_seed_bytes(&material))
}

pub fn derive_hash(password: &[u8], salt: &[u8], params: &Params) -> Result<Vec<u8>, Error> {
    if password.len() > MAX_PASSWORD {
        return Err(Error::Params("password too long"));
    }
    params.validate()?;
    if salt.len() < 16 || salt.len() > 64 {
        return Err(Error::Params("salt length out of range"));
    }

    let material = build_encoded_input(password, salt, params);
    let seed = derive_seed_bytes(&material);

    let mut memory = Memory::new(params.memory_kib, params.parallelism);
    initialize(&mut memory, &seed);
    fill(&mut memory, params.iterations);
    Ok(finalize(&memory, &seed, params))
}

fn initialize(memory: &mut Memory, seed: &[u8; 32]) {
    let mut expand_input = [0u8; 40];
    expand_input[..32].copy_from_slice(seed);
    let mut block_bytes = [0u8; BLOCK_SIZE];
    let mut words = [0u64; WORDS];

    for lane in 0..memory.parallelism {
        for block_index in 0..2 {
            expand_input[32..36].copy_from_slice(&le32(lane as u32));
            expand_input[36..40].copy_from_slice(&le32(block_index as u32));
            expand(&expand_input, &mut block_bytes);
            bytes_to_words(&block_bytes, &mut words);
            memory.block_mut(lane, block_index).copy_from_slice(&words);
        }
    }
}

fn fill(memory: &mut Memory, iterations: usize) {
    let mut combined = [0u64; WORDS];
    let mut mixed = [0u64; WORDS];
    let mut previous = [0u64; WORDS];
    let mut reference = [0u64; WORDS];
    let mut output = [0u64; WORDS];

    for pass in 0..iterations {
        for slice in 0..4 {
            for lane in 0..memory.parallelism {
                process_slice(
                    memory,
                    pass,
                    slice,
                    lane,
                    &mut combined,
                    &mut mixed,
                    &mut previous,
                    &mut reference,
                    &mut output,
                );
            }
        }
    }
}

fn process_slice(
    memory: &mut Memory,
    pass: usize,
    slice: usize,
    lane: usize,
    combined: &mut [u64; WORDS],
    mixed: &mut [u64; WORDS],
    previous: &mut [u64; WORDS],
    reference: &mut [u64; WORDS],
    output: &mut [u64; WORDS],
) {
    let mut start = slice * memory.slice_length;
    let end = start + memory.slice_length;
    if pass == 0 && slice == 0 {
        start = 2;
    }

    for block_index in start..end {
        let previous_index = if block_index > 0 {
            block_index - 1
        } else {
            memory.blocks_per_lane - 1
        };
        memory.copy_block(lane, previous_index, previous);

        let (ref_lane, ref_index) =
            select_reference(memory, pass, slice, lane, block_index, previous);
        memory.copy_block(ref_lane, ref_index, reference);
        xor_words(combined, previous, reference);

        if pass == 0 {
            mix(combined, pass as u64, lane as u64, block_index as u64, output);
            memory.block_mut(lane, block_index).copy_from_slice(output);
        } else {
            mix(combined, pass as u64, lane as u64, block_index as u64, mixed);
            memory.copy_block(lane, block_index, previous); // reuse as old-current scratch
            xor_words(output, previous, mixed);
            memory.block_mut(lane, block_index).copy_from_slice(output);
        }
    }
}

fn select_reference(
    memory: &Memory,
    pass: usize,
    slice: usize,
    current_lane: usize,
    block_index: usize,
    previous: &[u64; WORDS],
) -> (usize, usize) {
    let address_word = previous[0]
        ^ rotl(previous[17], 13)
        ^ previous[73]
        ^ (pass as u64)
        ^ rotl(block_index as u64, 29);

    let mut reference_lane = if memory.parallelism == 1 {
        0
    } else if block_index % 32 == 0 {
        fast_range(previous[1], memory.parallelism as u64) as usize
    } else {
        current_lane
    };

    let mut allowed = allowed_block_count(memory, pass, slice, current_lane, reference_lane, block_index);
    if allowed == 0 {
        reference_lane = current_lane;
        allowed = allowed_block_count(memory, pass, slice, current_lane, reference_lane, block_index);
    }
    assert!(allowed > 0);
    let reference_index = fast_range(address_word, allowed as u64) as usize;
    (reference_lane, reference_index)
}

fn allowed_block_count(
    memory: &Memory,
    pass: usize,
    slice: usize,
    current_lane: usize,
    reference_lane: usize,
    block_index: usize,
) -> usize {
    if reference_lane == current_lane {
        if pass == 0 {
            block_index
        } else {
            memory.blocks_per_lane
        }
    } else {
        slice * memory.slice_length
    }
}

fn finalize(memory: &Memory, seed: &[u8; 32], params: &Params) -> Vec<u8> {
    let mut acc = [0u64; WORDS];
    let last = memory.blocks_per_lane - 1;
    let quarter = memory.blocks_per_lane / 4;
    let half = memory.blocks_per_lane / 2;
    let three_quarter = (memory.blocks_per_lane * 3) / 4;
    let mut tmp = [0u64; WORDS];

    for lane in 0..memory.parallelism {
        for idx in [last, quarter, half, three_quarter] {
            memory.copy_block(lane, idx, &mut tmp);
            for i in 0..WORDS {
                acc[i] ^= tmp[i];
            }
        }
    }

    let mut accumulator = [0u8; BLOCK_SIZE];
    words_to_bytes(&acc, &mut accumulator);
    let group_root = compute_group_root(memory);

    let mut root_input = Vec::new();
    root_input.extend_from_slice(FINAL_PREFIX.as_bytes());
    root_input.extend_from_slice(seed);
    root_input.extend_from_slice(&accumulator);
    root_input.extend_from_slice(&group_root);
    root_input.extend_from_slice(&le32(params.memory_kib as u32));
    root_input.extend_from_slice(&le32(params.iterations as u32));
    root_input.extend_from_slice(&le32(params.parallelism as u32));
    root_input.extend_from_slice(&le32(params.output_length as u32));
    let root = blake3_hash(&root_input);

    let mut output_input = Vec::new();
    output_input.extend_from_slice(OUTPUT_PREFIX.as_bytes());
    output_input.extend_from_slice(&root);
    output_input.extend_from_slice(seed);
    blake3_xof(&output_input, params.output_length)
}

fn compute_group_root(memory: &Memory) -> [u8; 32] {
    let mut hasher = blake3::Hasher::new();
    hasher.update(GROUP_ROOT_PREFIX.as_bytes());

    const GROUP: usize = 64;
    let total = memory.parallelism * memory.blocks_per_lane;
    let group_count = (total + GROUP - 1) / GROUP;
    let mut group_buf = vec![0u8; GROUP * BLOCK_SIZE];
    let mut block_bytes = [0u8; BLOCK_SIZE];
    let mut words = [0u64; WORDS];

    for group_index in 0..group_count {
        let start = group_index * GROUP;
        let count = (total - start).min(GROUP);
        let byte_len = count * BLOCK_SIZE;
        for i in 0..count {
            let flat = start + i;
            let lane = flat / memory.blocks_per_lane;
            let block_index = flat % memory.blocks_per_lane;
            memory.copy_block(lane, block_index, &mut words);
            words_to_bytes(&words, &mut block_bytes);
            group_buf[i * BLOCK_SIZE..(i + 1) * BLOCK_SIZE].copy_from_slice(&block_bytes);
        }

        let mut group_hasher = blake3::Hasher::new();
        group_hasher.update(GROUP_PREFIX.as_bytes());
        group_hasher.update(&le64(group_index as u64));
        group_hasher.update(&le64(count as u64));
        group_hasher.update(&group_buf[..byte_len]);
        let digest = group_hasher.finalize();

        hasher.update(&le64(group_index as u64));
        hasher.update(digest.as_bytes());
    }

    *hasher.finalize().as_bytes()
}
